### PR TITLE
New Asset System - Move lazy migration into asset-router

### PR DESCRIPTION
**Pull request by sea212**
### What does it do?
- Removes "Old" and "New" market token variants from `Assets`, `Currencies` and `MarketAssets`.
- Applies proper routing for old and new assets in `asset-router`

### What important points should reviewers know?
- Not doing this would lead to convoluted structs and shifting the purpose of deciding what's old and new to the entity that invokes a dispatchable.

### Is there something left for follow-up PRs?

- Lazy migrate existing market assets: Pallets will continue using orml-tokens for existing market until all those existing market assets expired. Reason: Too many accounts hold existing market assets, migrating will be expensive and cannot be executed atomically. During migration, the storage trie that hold old market asset information will most likely be manipulated by user interaction.
- Paying fees using CampaignAssets.
- Using CampaignAssets and CustomAssets as base assets in prediction markets.
- Properly creating and destroying MarketAssets in pallets such as prediction markets.
- Adjust Asset types in many pallets, such as xcm related pallets (should only allow ZTG and ForeignAsset(id) for now), etc.
- Migrate WSX to CampaignAssets
- Future: Remove lazy migration logic once all old outcome tokens were redeemed

### What alternative implementations were considered?
- Leaving it as is, leads to convoluted structs and uncertainty about function parameters (i.e. user has to decide what is old or new).

### References
- Continuation of https://github.com/zeitgeistpm/zeitgeist/pull/1201

### CODE REVIEW

File: primitives/src/assets/market_assets.rs [hunk 1]
```diff
@@ -27,58 +27,34 @@ use super::*;
...
-    OldCategoricalOutcome(MI, CategoryIndex),
+    CategoricalOutcome(MI, CategoryIndex),
 
     #[codec(index = 1)]
-    OldScalarOutcome(MI, ScalarPosition),
+    ScalarOutcome(MI, ScalarPosition),
 
     #[codec(index = 3)]
-    OldPoolShare(PoolId),
+    PoolShare(PoolId),
```
Comment (Chralt98) [inline primitives/src/assets/market_assets.rs:37]: Just a general question: Wouldn't even renaming struct fields lead to storage migrations?

File: zrml/asset-router/src/pallet_impl/inspect.rs
Comment (Chralt98) [inline zrml/asset-router/src/pallet_impl/inspect.rs:1]: Why this change?

File: zrml/asset-router/src/pallet_impl/named_multi_reserveable_currency.rs [hunk 2]
```diff
@@ -91,7 +134,17 @@ impl<T: Config> NamedMultiReservableCurrency<T::AccountId> for Pallet<T> {
...
+            if T::MarketAssets::asset_exists(asset) {
+                return Err(Error::<T>::Unsupported.into());
+            }
+
+            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                return <T::Currencies as NamedMultiReservableCurrency<T::AccountId>>::repatriate_reserved_named(
+                    id, currency, slashed, beneficiary, value, status
+                );
+            }
+        } else if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
```
Comment (Chralt98) [inline zrml/asset-router/src/pallet_impl/named_multi_reserveable_currency.rs:147]: Instead of rewriting the same, why not instead of using `else if let Ok` we just use `if let Ok`?

Example:

```rust
    fn repatriate_reserved_named(
        id: &Self::ReserveIdentifier,
        currency_id: Self::CurrencyId,
        slashed: &T::AccountId,
        beneficiary: &T::AccountId,
        value: Self::Balance,
        status: Status,
    ) -> Result<Self::Balance, DispatchError> {
        if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
            if T::MarketAssets::asset_exists(asset) {
                return Err(Error::<T>::Unsupported.into());
            }
        }
        if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
            return <T::Currencies as NamedMultiReservableCurrency<T::AccountId>>::repatriate_reserved_named(
                id, currency, slashed, beneficiary, value, status
            );
        }

        Err(Error::<T>::Unsupported.into())
    }
```

You might want to do the same above.

**Review by Chralt98 — COMMENTED** [review]
This new asset system is pretty complex now, but I see why it is important. I sometimes ask myself, if it weren't been better developer deciding themselves when to use `pallet-assets` and when `orml-tokens`, but since you are already that committed and did consider this option too, I need to deal with this complexity now. But I also think, if the developer doesn't has to deal with when to use `orml-tokens` and when `pallet-assets` your abstraction is useful. But if we make a mistake inside the abstraction (asset-router), we need to dig into the complexity again. I shouldn't complain... It's a trade-off.

General question: Why didn't we avoid `New` and `Old` in the first place? I mean having `New` and `Old` struct fields was confusing back then too. What about storage migrations? I assume we are not using the new asset system in production at the moment and that's why it's not a storage disruption, right?

File: primitives/src/assets/market_assets.rs (hunk 1 above)
Comment (sea212) [reply primitives/src/assets/market_assets.rs:37]: Nope, they are simply encoded into `concat(index, type)`. You can see that types are properly encoded and decoded [within the asset scale_codec tests](https://github.com/zeitgeistpm/zeitgeist/blob/sea212-new-asset-system-part-5/primitives/src/assets/tests/scale_codec.rs)

File: zrml/asset-router/src/pallet_impl/inspect.rs
Comment (sea212) [reply zrml/asset-router/src/pallet_impl/inspect.rs:1]: It's necessary because multiple outcome types are mapped to both `Currencies` and `MarketAssets`. Checking if a "new" `MarketAsset` would be impossible with the previous code, because `asset_exists` will always return true because the outcome successfully converts into `Currencies` ("old" type). This is also causing issues the other way around, i.e. simply testing for `MarketAsset` first and then `Currencies` would lead to the same issue.

File: zrml/asset-router/src/pallet_impl/named_multi_reserveable_currency.rs (hunk 2 above)
Comment (sea212) [reply zrml/asset-router/src/pallet_impl/named_multi_reserveable_currency.rs]: https://github.com/zeitgeistpm/zeitgeist/pull/1256/commits/b5922a506c94b35d50c01a96ce803c58719281d8

**Comment by sea212** [comment]
> Why didn't we avoid New and Old in the first place? I mean having New and Old struct fields was confusing back then too.

@Chralt98 Initially the idea was to encapsulate the whole lazy migration into the scale codec magic. This was achieved by providing "New" and "Old" variants. Since there was always only a one to one mapping between any asset enum and the overarching `Assets` asset enum, there was no additional routing logic necessary to execute the lazy migration. Example:
1. `OrderBook` reads an `Order` from storage
2. The `Order` contains an old outcome
3. The old outcome is mapped to `Assets::Outcome...`
4. `Assets::Outcome...` is mapped to `Currencies::Outcome`

Thus, no additional logic is necessary to do the "lazy migration", as the encoding does all the proper mapping
This would even work when a pallet starts to assume `MarketAsset` types for outcomes, as an old outcome would have been mapped to `MarketAssets::OldOutcome...` which would have been mapped to `Assets::Outcome...` which in turn would have been mapped to `Currencies::Outcome`. On the other hand, reading a new market asset from storage would result in `MarketAssets::Outcome` which would have been mapped to `Assets::NewOutcome` which finally would have been mapped to `MarketAssets::Outcome` again. The advantage of this approach is that no additional logic for the lazy migration is necessary, but the unacceptable disadvantage is that whenever a dispatchable call uses `MarketAssets` or `Assets` as a type, the caller is responsible for determining whether the old or new variant has to be selected. The new solution solves that problem.

**Review by Chralt98 — APPROVED** [review]

File: zrml/asset-router/src/pallet_impl/multicurrency.rs [hunk 3]
```diff
@@ -46,14 +46,21 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
-        if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-            <T::Currencies as MultiCurrency<T::AccountId>>::free_balance(currency, who)
-        } else if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
-            T::MarketAssets::reducible_balance(asset, who, false)
+        if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
+            if T::MarketAssets::asset_exists(asset) {
+                T::MarketAssets::reducible_balance(asset, who, false)
+            } else if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                T::Currencies::free_balance(currency, who)
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/asset-router/src/pallet_impl/multicurrency.rs:54]: AFAIU, this is reached if `free_balance` is called with a market asset (meaning anything that successfully converts into a `T::MarketAssetType` object) which isn't registered. We seem to be missing this because we're using `ForeignAsset` for testing.

File: zrml/asset-router/src/pallet_impl/multicurrency.rs [hunk 4]
```diff
@@ -46,14 +46,21 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
+            } else {
+                T::MarketAssets::reducible_balance(asset, who, false)
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/asset-router/src/pallet_impl/multicurrency.rs:56]: AFAIU, this is unreachable as every market asset will successfully convert to a currency.

[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/asset-router/src/pallet_impl/multicurrency.rs:56]: If that's correct then I would opt for marking this as unreachable and return a dummy value.

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: zrml/asset-router/src/tests/destroy.rs [hunk 5]
```diff
@@ -21,12 +21,12 @@ use super::*;
 use frame_support::traits::tokens::fungibles::Inspect;
 
 fn test_helper(asset: Assets, initial_amount: <Runtime as crate::Config>::Balance) {
-    assert!(AssetRouter::asset_exists(asset));
     assert_ok!(<AssetRouter as orml_traits::MultiCurrency<AccountId>>::deposit(
         asset,
         &ALICE,
         initial_amount
     ));
+    assert!(AssetRouter::asset_exists(asset));
```
Comment (maltekliemann) [inline zrml/asset-router/src/tests/destroy.rs:29]: What's the purpose of this change?

File: zrml/asset-router/src/tests/inspect.rs [hunk 6]
```diff
@@ -21,12 +21,12 @@ use super::*;
 use frame_support::traits::tokens::fungibles::Inspect;
 
 fn test_helper(asset: Assets, initial_amount: <Runtime as crate::Config>::Balance) {
-    assert!(AssetRouter::asset_exists(asset));
     assert_ok!(<AssetRouter as orml_traits::MultiCurrency<AccountId>>::deposit(
         asset,
         &ALICE,
         initial_amount
     ));
+    assert!(AssetRouter::asset_exists(asset));
```
Comment (maltekliemann) [inline zrml/asset-router/src/tests/inspect.rs:29]: What's the purpose of this change?

File: zrml/asset-router/src/tests/mod.rs [hunk 7]
```diff
@@ -41,7 +41,7 @@ mod custom_types;
 mod destroy;
 mod inspect;
 mod managed_destroy;
-mod multi_currency;
 mod multi_lockable_currency;
 mod multi_reservable_currency;
+mod multicurrency;
```
Comment (maltekliemann) [inline zrml/asset-router/src/tests/mod.rs:46]: Why did we rename this? Seems more appropriate to call it `multi_currency` given that you're calling it `MultiCurrency` instead of `Multicurrency`.

File: runtime/battery-station/src/parameters.rs [hunk 8]
```diff
@@ -508,10 +508,10 @@ parameter_type_with_key! {
...
     pub ExistentialDeposits: |currency_id: Currencies| -> Balance {
         match currency_id {
-            Currencies::OldCategoricalOutcome(_,_) => ExistentialDeposit::get(),
-            Currencies::OldParimutuelShare(_,_)  => ExistentialDeposit::get(),
-            Currencies::OldPoolShare(_)  => ExistentialDeposit::get(),
-            Currencies::OldScalarOutcome(_,_)  => ExistentialDeposit::get(),
+            Currencies::CategoricalOutcome(_,_) => ExistentialDeposit::get(),
+            Currencies::ParimutuelShare(_,_)  => ExistentialDeposit::get(),
+            Currencies::PoolShare(_)  => ExistentialDeposit::get(),
+            Currencies::ScalarOutcome(_,_)  => ExistentialDeposit::get(),
```
[the remark this answers is not in the document]
Comment (sea212) [reply runtime/battery-station/src/parameters.rs]: https://github.com/zeitgeistpm/zeitgeist/pull/1256/commits/a0ea772952c50cdd47535ac5b212a383f0e34b3a

File: runtime/zeitgeist/src/parameters.rs [hunk 9]
```diff
@@ -506,10 +506,10 @@ parameter_type_with_key! {
...
     pub ExistentialDeposits: |currency_id: Currencies| -> Balance {
         match currency_id {
-            Currencies::OldCategoricalOutcome(_,_) => ExistentialDeposit::get(),
-            Currencies::OldParimutuelShare(_,_)  => ExistentialDeposit::get(),
-            Currencies::OldPoolShare(_)  => ExistentialDeposit::get(),
-            Currencies::OldScalarOutcome(_,_)  => ExistentialDeposit::get(),
+            Currencies::CategoricalOutcome(_,_) => ExistentialDeposit::get(),
+            Currencies::ParimutuelShare(_,_)  => ExistentialDeposit::get(),
+            Currencies::PoolShare(_)  => ExistentialDeposit::get(),
+            Currencies::ScalarOutcome(_,_)  => ExistentialDeposit::get(),
```
[the remark this answers is not in the document]
Comment (sea212) [reply runtime/zeitgeist/src/parameters.rs]: https://github.com/zeitgeistpm/zeitgeist/pull/1256/commits/f0f88e97d2f607fb8501f5bb7b17e55c8e7b0481

File: zrml/asset-router/src/tests/mod.rs (hunk 7 above)
Comment (sea212) [reply zrml/asset-router/src/tests/mod.rs]: https://github.com/zeitgeistpm/zeitgeist/pull/1256/commits/2134e4a76ac9fbd07107baa10895a44c552e6c03

File: zrml/asset-router/src/tests/inspect.rs (hunk 6 above)
Comment (sea212) [reply zrml/asset-router/src/tests/inspect.rs:29]: If that assertion is invoked before the deposit, the `Currencies` case will error, because in the `Currencies` case the asset is considered as existing when `total_issuance > 0`.

File: zrml/asset-router/src/tests/destroy.rs (hunk 5 above)
Comment (sea212) [reply zrml/asset-router/src/tests/destroy.rs:29]: See https://github.com/zeitgeistpm/zeitgeist/pull/1256#discussion_r1497273125

File: zrml/asset-router/src/macros.rs [hunk 10]
```diff
@@ -37,7 +46,21 @@ macro_rules! route_call {
 /// it returns an error.
 macro_rules! only_currency {
     ($currency_id:expr, $error:expr, $currency_trait:ident, $currency_method:ident, $($args:expr),+) => {
-        if let Ok(currency) = T::CurrencyType::try_from($currency_id) {
+        if let Ok(asset) = T::MarketAssetType::try_from($currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
+            if T::MarketAssets::asset_exists(asset) {
+                Self::log_unsupported($currency_id, stringify!($currency_method));
```
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/macros.rs:52]: Yeah, if it is invoked in the context where `$currency_method` is valid for a `MarketAsset` (or strictly speaking anything but `Currencies`), then it's reachable.

File: zrml/asset-router/src/macros.rs [hunk 11]
```diff
@@ -37,7 +46,21 @@ macro_rules! route_call {
...
+                $error
+            } else {
+                if let Ok(currency) = T::CurrencyType::try_from($currency_id) {
+                    <T::Currencies as $currency_trait<T::AccountId>>::$currency_method(currency, $($args),+)
+                } else {
+                    Self::log_unsupported($currency_id, stringify!($currency_method));
```
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/macros.rs:58]: Should be

File: zrml/asset-router/src/pallet_impl/multicurrency.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/pallet_impl/multicurrency.rs:54]: https://github.com/zeitgeistpm/zeitgeist/pull/1256/commits/9df31a5733ef97dd3b1f53be6c92ab9a22adc636

File: zrml/asset-router/src/pallet_impl/multicurrency.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/pallet_impl/multicurrency.rs:56]: This is true under the assumption that all asset types within `MarketAsset` are contained within `Currencies` (as you stated). Should we decide to add another asset type to `MarketAsset` in the future while the lazy migration still is not completed, the code would jump into the dummy value instead of executing the function in `MarketAssets`. For this reason I tend to leave it as is.

File: zrml/asset-router/src/pallet_impl/multicurrency.rs [hunk 12]
```diff
@@ -116,30 +133,46 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
...
     fn can_slash(currency_id: Self::CurrencyId, who: &T::AccountId, value: Self::Balance) -> bool {
-        if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-            <T::Currencies as MultiCurrency<T::AccountId>>::can_slash(currency, who, value)
-        } else if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
-            T::MarketAssets::reducible_balance(asset, who, false) >= value
+        if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
+            if T::MarketAssets::asset_exists(asset) {
+                // Resulting balance can be ignored as `burn_from` ensures that the
+                // requested amount can be burned.
```
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/pallet_impl/multicurrency.rs:163]: This comment is incorrect.

it invokes [`do_burn`](https://github.com/paritytech/substrate/blob/bcff60a227d455d95b4712b6cb356ce56b1ff672/frame/assets/src/impl_fungibles.rs#L101-L102)

**Note that `best_effort: false`.**

it invokes [`decrease_balance`](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L497)

it invokes [`prep_debit`](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L537)

[`prep_debit` docstring](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L249-L253) clearly states that `best_effort` determines whether less than `amount` shall be accepted. If `false`, then no. It can also be verified at the [respective `ensure!` call](https://github.com/paritytech/substrate/blob/033d4e86cc7eff0066cd376b9375f815761d653c/frame/assets/src/functions.rs#L263)

File: zrml/asset-router/src/pallet_impl/multicurrency_extended.rs [hunk 13]
```diff
@@ -25,29 +54,27 @@ impl<T: Config> MultiCurrencyExtended<T::AccountId> for Pallet<T> {
...
-            Self::deposit(currency_id, who, by_balance)
+        if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
+            if T::MarketAssets::asset_exists(asset) {
+                Self::update_balance_asset(currency_id, who, by_amount)
+            } else if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                T::Currencies::update_balance(currency, who, by_amount)
+            } else {
+                Self::update_balance_asset(currency_id, who, by_amount)
+            }
```
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/pallet_impl/multicurrency_extended.rs]: I don't think that it is necessary. In addition to https://github.com/zeitgeistpm/zeitgeist/pull/1256#discussion_r1497376978 (which shows that it should not be assumed that it's `unreachable`), the invocation of `deposit` or `withdraw` will provide an appropriate errors that the asset does not exist (not created).

File: zrml/asset-router/src/macros.rs [hunk 14]
```diff
@@ -19,14 +19,23 @@
 /// implementation that handles it and finally calls the $method on it.
 macro_rules! route_call {
     ($currency_id:expr, $currency_method:ident, $asset_method:ident, $($args:expr),*) => {
-        if let Ok(currency) = T::CurrencyType::try_from($currency_id) {
-            Ok(<T::Currencies as MultiCurrency<T::AccountId>>::$currency_method(currency, $($args),*))
-        } else if let Ok(asset) = T::MarketAssetType::try_from($currency_id) {
-            Ok(T::MarketAssets::$asset_method(asset, $($args),*))
+        if let Ok(asset) = T::MarketAssetType::try_from($currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
```
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/asset-router/src/macros.rs:23]: https://github.com/zeitgeistpm/zeitgeist/pull/1256/commits/8fdd8b6b543075fc924be7b6bb2a052606495939

File: zrml/asset-router/src/pallet_impl/multicurrency.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/asset-router/src/pallet_impl/multicurrency.rs:56]: I see. Yeah, makes sense.

File: zrml/asset-router/src/macros.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/asset-router/src/macros.rs:58]: I think we should mark this as unreachable in the feature branch then. Nothing we can do here though,  I guess.

**Review by maltekliemann — APPROVED** [review]
